// @hubspot/api-client/lib/src/client.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const bottleneck_1 = __importDefault(require("bottleneck"));
const _ = __importStar(require("lodash"));
const qs = __importStar(require("querystring"));
const request = require("request");
// @ts-ignore
const pJson = __importStar(require("../../package.json"));
const api_1 = require("../codegen/crm/associations/api");
const api_2 = require("../codegen/crm/companies/api");
const api_3 = require("../codegen/crm/contacts/api");
const api_4 = require("../codegen/crm/deals/api");
const api_5 = require("../codegen/crm/extensions/cards/api");
const coreApi_1 = require("../codegen/crm/imports/api/coreApi");
const api_6 = require("../codegen/crm/line_items/api");
const api_7 = require("../codegen/crm/owners/api");
const api_8 = require("../codegen/crm/pipelines/api");
const api_9 = require("../codegen/crm/products/api");
const api_10 = require("../codegen/crm/properties/api");
const api_11 = require("../codegen/crm/quotes/api");
const api_12 = require("../codegen/crm/tickets/api");
const api_13 = require("../codegen/oauth/api");
const DEFAULT_HEADERS = { 'User-Agent': `${pJson.name};${pJson.version}` };
const DEFAULT_LIMITER_OPTIONS = {
    minTime: 1000 / 9,
    maxConcurrent: 5,
};
const METHOD_NAMES_TO_EXCLUDE_FROM_PATCHING = [
    'constructor',
    'useQuerystring',
    'basePath',
    'defaultHeaders',
    'setDefaultAuthentication',
    'setApiKey',
    'accessToken',
    'addInterceptor',
];
const RETRY_TIMEOUT = {
    INTERNAL_SERVER_ERROR: 200,
    TOO_MANY_REQUESTS: 10 * 1000
};
const TEN_SECONDLY_ROLLING = "TEN_SECONDLY_ROLLING";
var NumberOfRetries;
(function (NumberOfRetries) {
    NumberOfRetries[NumberOfRetries["NoRetries"] = 0] = "NoRetries";
    NumberOfRetries[NumberOfRetries["One"] = 1] = "One";
    NumberOfRetries[NumberOfRetries["Two"] = 2] = "Two";
    NumberOfRetries[NumberOfRetries["Three"] = 3] = "Three";
    NumberOfRetries[NumberOfRetries["Four"] = 4] = "Four";
    NumberOfRetries[NumberOfRetries["Five"] = 5] = "Five";
    NumberOfRetries[NumberOfRetries["Six"] = 6] = "Six";
})(NumberOfRetries = exports.NumberOfRetries || (exports.NumberOfRetries = {}));
class HttpError extends Error {
    constructor(response, body, statusCode) {
        super('HTTP request failed');
        this.response = response;
        this.body = body;
        this.statusCode = statusCode;
        this.name = 'HttpError';
    }
}
exports.HttpError = HttpError;
class Client {
    constructor(options = {}) {
        this._interceptors = [];
        this._basePath = 'https://api.hubapi.com';
        this.authentications = {
            'hapikey': new api_7.ApiKeyAuth('query', 'hapikey'),
            'oauth2': new api_7.OAuth(),
        };
        this._useLimiter = true;
        this._oauthDefaultApi = new api_13.DefaultApi();
        this._associationsBatchApi = new api_1.BatchApi();
        this._companiesAssociationsApi = new api_2.AssociationsApi();
        this._companiesBasicApi = new api_2.BasicApi();
        this._companiesBatchApi = new api_2.BatchApi();
        this._companiesSearchApi = new api_2.SearchApi();
        this._contactsAssociationsApi = new api_3.AssociationsApi();
        this._contactsBasicApi = new api_3.BasicApi();
        this._contactsBatchApi = new api_3.BatchApi();
        this._contactsSearchApi = new api_3.SearchApi();
        this._dealsAssociationsApi = new api_4.AssociationsApi();
        this._dealsBasicApi = new api_4.BasicApi();
        this._dealsBatchApi = new api_4.BatchApi();
        this._dealsSearchApi = new api_4.SearchApi();
        this._cardsApi = new api_5.CardsApi();
        this._importsCoreApi = new coreApi_1.CoreApi();
        this._lineItemsAssociationsApi = new api_6.AssociationsApi();
        this._lineItemsBasicApi = new api_6.BasicApi();
        this._lineItemsBatchApi = new api_6.BatchApi();
        this._lineItemsSearchApi = new api_6.SearchApi();
        this._ownersDefaultApi = new api_7.DefaultApi();
        this._pipelinesApi = new api_8.PipelinesApi();
        this._pipelineStagesApi = new api_8.PipelineStagesApi();
        this._productsAssociationsApi = new api_9.AssociationsApi();
        this._productsBasicApi = new api_9.BasicApi();
        this._productsBatchApi = new api_9.BatchApi();
        this._productsSearchApi = new api_9.SearchApi();
        this._propertiesBatchApi = new api_10.BatchApi();
        this._propertiesCoreApi = new api_10.CoreApi();
        this._propertiesGroupsApi = new api_10.GroupsApi();
        this._quotesAssociationsApi = new api_11.AssociationsApi();
        this._quotesBasicApi = new api_11.BasicApi();
        this._quotesBatchApi = new api_11.BatchApi();
        this._quotesSearchApi = new api_11.SearchApi();
        this._ticketsAssociationsApi = new api_12.AssociationsApi();
        this._ticketsBasicApi = new api_12.BasicApi();
        this._ticketsBatchApi = new api_12.BatchApi();
        this._ticketsSearchApi = new api_12.SearchApi();
        this._apiClientsWithAuth = [
            this._associationsBatchApi,
            this._companiesAssociationsApi,
            this._companiesBasicApi,
            this._companiesBatchApi,
            this._companiesSearchApi,
            this._contactsAssociationsApi,
            this._contactsBasicApi,
            this._contactsBatchApi,
            this._contactsSearchApi,
            this._dealsAssociationsApi,
            this._dealsBasicApi,
            this._dealsBatchApi,
            this._dealsSearchApi,
            this._cardsApi,
            this._importsCoreApi,
            this._lineItemsAssociationsApi,
            this._lineItemsBasicApi,
            this._lineItemsBatchApi,
            this._lineItemsSearchApi,
            this._ownersDefaultApi,
            this._pipelinesApi,
            this._pipelineStagesApi,
            this._productsAssociationsApi,
            this._productsBasicApi,
            this._productsBatchApi,
            this._productsSearchApi,
            this._propertiesBatchApi,
            this._propertiesCoreApi,
            this._propertiesGroupsApi,
            this._quotesAssociationsApi,
            this._quotesBasicApi,
            this._quotesBatchApi,
            this._quotesSearchApi,
            this._ticketsAssociationsApi,
            this._ticketsBasicApi,
            this._ticketsBatchApi,
            this._ticketsSearchApi,
        ];
        this._apiClients = this._apiClientsWithAuth.slice();
        this._apiClients.push(this._oauthDefaultApi);
        this._numberOfApiCallRetries = NumberOfRetries.NoRetries;
        this._setUseQuerystring(true);
        this._setOptions(options);
        this.crm = {
            associations: {
                batchApi: this._associationsBatchApi,
            },
            companies: {
                associationsApi: this._companiesAssociationsApi,
                basicApi: this._companiesBasicApi,
                batchApi: this._companiesBatchApi,
                searchApi: this._companiesSearchApi,
            },
            contacts: {
                associationsApi: this._contactsAssociationsApi,
                basicApi: this._contactsBasicApi,
                batchApi: this._contactsBatchApi,
                searchApi: this._contactsSearchApi,
            },
            deals: {
                associationsApi: this._dealsAssociationsApi,
                basicApi: this._dealsBasicApi,
                batchApi: this._dealsBatchApi,
                searchApi: this._dealsSearchApi,
            },
            extensions: {
                cards: {
                    cardsApi: this._cardsApi,
                },
            },
            imports: {
                coreApi: this._importsCoreApi,
            },
            lineItems: {
                associationsApi: this._lineItemsAssociationsApi,
                basicApi: this._lineItemsBasicApi,
                batchApi: this._lineItemsBatchApi,
                searchApi: this._lineItemsSearchApi,
            },
            owners: {
                defaultApi: this._ownersDefaultApi,
            },
            pipelines: {
                pipelinesApi: this._pipelinesApi,
                pipelineStagesApi: this._pipelineStagesApi,
            },
            products: {
                associationsApi: this._productsAssociationsApi,
                basicApi: this._productsBasicApi,
                batchApi: this._productsBatchApi,
                searchApi: this._productsSearchApi,
            },
            properties: {
                batchApi: this._propertiesBatchApi,
                coreApi: this._propertiesCoreApi,
                groupsApi: this._propertiesGroupsApi,
            },
            quotes: {
                associationsApi: this._quotesAssociationsApi,
                basicApi: this._quotesBasicApi,
                batchApi: this._quotesBatchApi,
                searchApi: this._quotesSearchApi,
            },
            tickets: {
                associationsApi: this._ticketsAssociationsApi,
                basicApi: this._ticketsBasicApi,
                batchApi: this._ticketsBatchApi,
                searchApi: this._ticketsSearchApi,
            },
        };
        this.oauth = {
            defaultApi: this._oauthDefaultApi,
            getAuthorizationUrl: this._getAuthorizationUrl.bind(this),
        };
    }
    setApiKey(apiKeyToSet) {
        this._apiKey = apiKeyToSet;
        this.authentications.hapikey.apiKey = apiKeyToSet;
        _.each(this._apiClientsWithAuth, (apiClient) => {
            apiClient.setApiKey(0, apiKeyToSet);
        });
    }
    setBasePath(basePathToSet) {
        if (_.isNil(basePathToSet)) {
            return;
        }
        this._basePath = basePathToSet;
        const oauthBasePath = `${basePathToSet}/oauth`.replace(/\/+$/, '');
        const associationsBasePath = `${basePathToSet}/crm/v3/associations`.replace(/\/+$/, '');
        const companiesBasePath = `${basePathToSet}/crm/v3/objects`.replace(/\/+$/, '');
        const contactsBasePath = `${basePathToSet}/crm/v3/objects`.replace(/\/+$/, '');
        const dealsBasePath = `${basePathToSet}/crm/v3/objects`.replace(/\/+$/, '');
        const cardsBasePath = `${basePathToSet}/crm/v3/extensions/cards`.replace(/\/+$/, '');
        const importsBasePath = `${basePathToSet}/crm/v3/imports`.replace(/\/+$/, '');
        const lineItemsBasePath = `${basePathToSet}/crm/v3/objects`.replace(/\/+$/, '');
        const ownersBasePath = `${basePathToSet}/crm/v3/owners`.replace(/\/+$/, '');
        const pipelinesBasePath = `${basePathToSet}/crm/v3/pipelines`.replace(/\/+$/, '');
        const productsBasePath = `${basePathToSet}/crm/v3/objects`.replace(/\/+$/, '');
        const propertiesBasePath = `${basePathToSet}/crm/v3/properties`.replace(/\/+$/, '');
        const quotesBasePath = `${basePathToSet}/crm/v3/objects`.replace(/\/+$/, '');
        const ticketsBasePath = `${basePathToSet}/crm/v3/objects`.replace(/\/+$/, '');
        this._oauthDefaultApi.basePath = oauthBasePath;
        this._associationsBatchApi.basePath = associationsBasePath;
        this._companiesAssociationsApi.basePath = companiesBasePath;
        this._companiesBasicApi.basePath = companiesBasePath;
        this._companiesBatchApi.basePath = companiesBasePath;
        this._companiesSearchApi.basePath = companiesBasePath;
        this._contactsAssociationsApi.basePath = contactsBasePath;
        this._contactsBasicApi.basePath = contactsBasePath;
        this._contactsBatchApi.basePath = contactsBasePath;
        this._contactsSearchApi.basePath = contactsBasePath;
        this._dealsAssociationsApi.basePath = dealsBasePath;
        this._dealsBasicApi.basePath = dealsBasePath;
        this._dealsBatchApi.basePath = dealsBasePath;
        this._dealsSearchApi.basePath = dealsBasePath;
        this._cardsApi.basePath = cardsBasePath;
        this._importsCoreApi.basePath = importsBasePath;
        this._lineItemsAssociationsApi.basePath = lineItemsBasePath;
        this._lineItemsBasicApi.basePath = lineItemsBasePath;
        this._lineItemsBatchApi.basePath = lineItemsBasePath;
        this._lineItemsSearchApi.basePath = lineItemsBasePath;
        this._ownersDefaultApi.basePath = ownersBasePath;
        this._pipelinesApi.basePath = pipelinesBasePath;
        this._pipelineStagesApi.basePath = pipelinesBasePath;
        this._productsAssociationsApi.basePath = productsBasePath;
        this._productsBasicApi.basePath = productsBasePath;
        this._productsBatchApi.basePath = productsBasePath;
        this._productsSearchApi.basePath = productsBasePath;
        this._propertiesBatchApi.basePath = propertiesBasePath;
        this._propertiesCoreApi.basePath = propertiesBasePath;
        this._propertiesGroupsApi.basePath = propertiesBasePath;
        this._quotesAssociationsApi.basePath = quotesBasePath;
        this._quotesBasicApi.basePath = quotesBasePath;
        this._quotesBatchApi.basePath = quotesBasePath;
        this._quotesSearchApi.basePath = quotesBasePath;
        this._ticketsAssociationsApi.basePath = ticketsBasePath;
        this._ticketsBasicApi.basePath = ticketsBasePath;
        this._ticketsBatchApi.basePath = ticketsBasePath;
        this._ticketsSearchApi.basePath = ticketsBasePath;
    }
    setAccessToken(accessTokenToSet) {
        this._accessToken = accessTokenToSet;
        this.authentications.oauth2.accessToken = accessTokenToSet;
        _.each(this._apiClientsWithAuth, (apiClient) => {
            apiClient.accessToken = accessTokenToSet;
        });
    }
    addInterceptor(interceptor) {
        this._interceptors.push(interceptor);
        _.each(this._apiClients, (apiClient) => {
            apiClient.addInterceptor(interceptor);
        });
    }
    setDefaultHeaders(defaultHeadersToSet) {
        this._defaultHeaders = _.assign({}, defaultHeadersToSet, DEFAULT_HEADERS);
        _.each(this._apiClients, (apiClient) => {
            apiClient.defaultHeaders = this._defaultHeaders;
        });
    }
    setAuth(options = {}) {
        if (options.apiKey) {
            this.setApiKey(options.apiKey);
        }
        else if (options.accessToken) {
            this.setAccessToken(options.accessToken);
        }
    }
    getOptions() {
        return {
            accessToken: this._accessToken,
            apiKey: this._apiKey,
            basePath: this._basePath,
            defaultHeaders: this._defaultHeaders,
            useLimiter: this._useLimiter,
            limiterOptions: this._limiterOptions,
            numberOfApiCallRetries: this._numberOfApiCallRetries,
            interceptors: this._interceptors,
        };
    }
    apiRequest(opts) {
        const params = _.cloneDeep(opts);
        params.method = params.method || 'GET';
        params.json = true;
        params.resolveWithFullResponse = true;
        params.url = params.overlapUrl || this._basePath + (params.path || '');
        delete params.overlapUrl;
        delete params.path;
        params.qsStringifyOptions = {
            arrayFormat: 'repeat',
        };
        params.qs = Object.assign({}, params.qs);
        params.headers = _.assign({}, opts.headers, this._defaultHeaders);
        if (this.authentications.hapikey.apiKey) {
            this.authentications.hapikey.applyToRequest(params);
        }
        if (this.authentications.oauth2.accessToken) {
            this.authentications.oauth2.applyToRequest(params);
        }
        let authenticationPromise = Promise.resolve();
        if (this.authentications.hapikey.apiKey) {
            authenticationPromise = authenticationPromise.then(() => this.authentications.hapikey.applyToRequest(params));
        }
        if (this.authentications.oauth2.accessToken) {
            authenticationPromise = authenticationPromise.then(() => this.authentications.oauth2.applyToRequest(params));
        }
        let interceptorPromise = authenticationPromise;
        for (const interceptor of this._interceptors) {
            interceptorPromise = interceptorPromise.then(() => interceptor(params));
        }
        return interceptorPromise.then(() => {
            return new Promise((resolve, reject) => {
                request(params, (error, response, body) => {
                    if (error) {
                        reject(error);
                    }
                    else {
                        if (response.statusCode && response.statusCode >= 200 && response.statusCode <= 299) {
                            resolve({ response, body });
                        }
                        else {
                            reject(new HttpError(response, body, response.statusCode));
                        }
                    }
                });
            });
        });
    }
    _getAuthorizationUrl(clientId, redirectUri, scopes) {
        const params = {
            client_id: clientId,
            redirect_uri: redirectUri,
            scopes,
        };
        return `https://app.hubspot.com/oauth/authorize?${qs.stringify(params)}`;
    }
    _setOptions(options) {
        this.setAuth(options);
        this.setBasePath(options.basePath);
        this.setDefaultHeaders(options.defaultHeaders);
        this._setMethodsPatchOptions(options);
        this._setInterceptors(options);
    }
    _setUseQuerystring(useQuerystring) {
        _.each(this._apiClients, (apiClient) => apiClient._useQuerystring = useQuerystring);
    }
    _getLimiterWrappedMethod(method) {
        if (!this._limiter) {
            throw new Error('Limiter not defined');
        }
        return this._limiter.wrap(method);
    }
    _waitAfterRequestFailure(statusCode, retryNumber, retryTimeout) {
        console.error(`Request failed with status code [${statusCode}], will retry [${retryNumber}] time in [${retryTimeout}] ms`);
        return new Promise((resolve) => setTimeout(resolve, retryTimeout * retryNumber));
    }
    _getRetryWrappedMethod(method) {
        return (...args) => __awaiter(this, void 0, void 0, function* () {
            const numberOfRetries = this._numberOfApiCallRetries.valueOf() + 1;
            let resultSuccess;
            let resultRejected;
            for (let index = 1; index <= numberOfRetries; index++) {
                try {
                    resultSuccess = yield method(...args);
                    resultRejected = null;
                    break;
                }
                catch (e) {
                    resultRejected = e;
                    if (_.isEqual(index, numberOfRetries)) {
                        break;
                    }
                    const statusCode = _.get(e, 'response.statusCode');
                    if (statusCode >= 500 && statusCode <= 599) {
                        yield this._waitAfterRequestFailure(statusCode, index, RETRY_TIMEOUT.INTERNAL_SERVER_ERROR);
                        continue;
                    }
                    if (_.isEqual(statusCode, 429)) {
                        const policyName = _.get(e, 'response.body.policyName');
                        if (_.isEqual(policyName, TEN_SECONDLY_ROLLING)) {
                            yield this._waitAfterRequestFailure(statusCode, index, RETRY_TIMEOUT.TOO_MANY_REQUESTS);
                            continue;
                        }
                    }
                    break;
                }
            }
            return new Promise((resolve, reject) => {
                if (resultRejected) {
                    return reject(resultRejected);
                }
                return resolve(resultSuccess);
            });
        });
    }
    _patchApiClientMethod(methodName, clientInstance, clientPrototype) {
        const methodToPatch = clientPrototype[methodName].bind(clientInstance);
        let patchedMethod = methodToPatch;
        if (this._useLimiter) {
            patchedMethod = this._getLimiterWrappedMethod(methodToPatch);
        }
        if (!_.isEqual(this._numberOfApiCallRetries, NumberOfRetries.NoRetries)) {
            patchedMethod = this._getRetryWrappedMethod(patchedMethod);
        }
        clientInstance[methodName] = patchedMethod;
    }
    _patchApiClient(clientInstance) {
        const clientPrototype = Object.getPrototypeOf(clientInstance);
        let methodsNamesToPatch = Object.getOwnPropertyNames(clientPrototype);
        methodsNamesToPatch = _.differenceWith(methodsNamesToPatch, METHOD_NAMES_TO_EXCLUDE_FROM_PATCHING);
        const patchFn = this._patchApiClientMethod.bind(this);
        _.each(methodsNamesToPatch, (methodName) => {
            patchFn(methodName, clientInstance, clientPrototype);
        });
    }
    _patchApiClients() {
        _.each(this._apiClients, this._patchApiClient.bind(this));
    }
    _patchApiRequestMethod() {
        let apiRequestMethodToPatch = this.apiRequest.bind(this);
        if (this._useLimiter) {
            apiRequestMethodToPatch = this._getLimiterWrappedMethod(apiRequestMethodToPatch);
        }
        if (!_.isEqual(this._numberOfApiCallRetries, NumberOfRetries.NoRetries)) {
            apiRequestMethodToPatch = this._getRetryWrappedMethod(apiRequestMethodToPatch);
        }
        this.apiRequest = apiRequestMethodToPatch;
    }
    _setMethodsPatchOptions(options = {}) {
        this._useLimiter = _.isNil(options.useLimiter) ? true : options.useLimiter;
        this._numberOfApiCallRetries = _.isNil(options.numberOfApiCallRetries) ? NumberOfRetries.NoRetries : options.numberOfApiCallRetries;
        if (this._useLimiter) {
            this._limiterOptions = _.isNil(options.limiterOptions) ? DEFAULT_LIMITER_OPTIONS : options.limiterOptions;
            this._limiter = new bottleneck_1.default(this._limiterOptions);
        }
        if (this._useLimiter || !_.isEqual(this._numberOfApiCallRetries, NumberOfRetries.NoRetries)) {
            this._patchApiClients();
            this._patchApiRequestMethod();
        }
    }
    _setInterceptors(options = {}) {
        if (options.interceptors) {
            _.each(options.interceptors, this.addInterceptor.bind(this));
        }
    }
}
exports.Client = Client;
